### ISSUE TITLE
[selectors] Add :role() pseudo-class

### ISSUE THREAD

**Issue by tabatkins** [w3c/csswg-drafts#3596]
In <https://lists.w3.org/Archives/Public/www-style/2015Jan/0004.html> (Oct 2014 Santa Clara f2f), we resolved to add a `:role()` pseudo-class, to match based on ARIA roles easily:

<details>
<summary>minutes excerpt</summary>
<pre>
TabAtkins: jcraig suggested a :role pseudo
hober: There's a few problems. One is the impl RA roles on the
elements. When you have multiple roles on one element, you
get all of them.
hober: For styling you may want to style that as a switch or a
checkbox. Right now that's difficult in CSS.
hober: There's also the case of wanting to simply style all the
buttons on the page.
hober: You want to be able to say :role(button) and be done.

TabAtkins: That's convincing. Is there anything outside that?
hober: We can always add more.

glazou: You mentioned implied role. So some won't need an assigned
role. We'll need to refer to HTML.
hober: In CSS we say the host language has that.

dbaron: It's :: or : ?
TabAtkins: It's a functional class. It should be an ident.
glazou: Can you have multiple roles?
hober: Syntactically you can,
hober: But only one is active.
glazou: So if you have multiple roles on one element, it's
whitespace broken?
dino: So if you do :role...
hober: You can do role=button and role="super button". Or we
separate in the list itself.
glazou: So the role attr is multiple but the :role() pseudo is only
one?
hober: Yeah.

dauwhe: dpub is working on adding things for role attr.
TabAtkins: Do you know if it will be consistent with this?
dauwhe: They haven't said. I'll raise that with them.

TabAtkins: So I guess that's simple. Objections to adding :role()
to selectors?
TabAtkins: So :role(button) would target anything that has a
button according to aria rules.
hober: It wouldn't match button role = checkbox or something crazy

dbaron: So people are okay with agreeing CSS won't do future
things to influence what the aria role is.
hober: It should be defined by pointing over there [to markup layer].
fantasai: We need to have another language that has CSS syntax for
that kind of stuff. Our stuff should be about styling.
fantasai: People keep asking us to put things in CSS that belong
in the DOM, because they want to use CSS selectors and
cascading to attach it to the tree.
TabAtkins: If someone would like to implement cascading attr
selectors in a browser shrug
dbaron: Having CSS do this side makes sense, but we can't do both.
hober: Should we have a note in the spec saying "you may be
confused"?
glazou: When we had selectors we didn't have a note. I'm not sure
we need a note.
hober: Okay.

glazou: So no objections?

RESOLVED: add :role() to selectors

glazou: Who is going to write the prose?
TabAtkins: I will.
</pre>
</details>

I volunteered to write the text, but clearly never did. ^_^

**Comment by bkardell** [comment]
I really like the idea of this and was thus going to take a stab at writing some proposed prose here, but .... from my recollection, and the minutes (both of which may be flawed), it feels like the 'key' things mentioned here really hinge on this being based off the computed role - like what is the actual role that this is playing.  But, in discussing this with @alice, this might actually be pretty tricky because what the computed role is sometimes involves fairly complex checking and tree walking.

**Comment by alice** [comment]
I'd be interested to know what use cases @cookiecrook had in mind here. 

I agree with @bkardell, it seems to be implied that this refers to computed ARIA-equivalent role (i.e. a `<button>` would match `:role(button)`, but that is not clear from the minutes.

Assuming that to be the case, implementing this could be quite complex.

**Comment by cookiecrook** [comment]
That is the case, but Alice is correct that implementation proved significantly more challenging than expected. 

Style parsing needs to happen before layout and rendering. In WebKit (and Blink, maybe more) the accessibility tree is computed based on the render tree. Since accessibility can change the roles (or refuse to change the roles of malformed nodes), current implementations would either result in one of the following:

1. A circular dependency (neither style nor accessibility can proceed until the other is complete)

2. CSS parsing continues but later changes when accessibility loads could trigger new layout and repaints.

We left off that path by determining that a non-trivial amount of WebCore/Accessibility would need to moved out of the accessibility code and refactored to be used prior to layout.

It’s not impossible, but it’s quite a bit more than anyone was willing to commit to.

A WICG incubator may be a good path forward for this one. You’ll need implementor involvement and buy-in before the :role() selector can proceed.

**Comment by cookiecrook** [comment]
@bkardell wrote:
> the computed role is sometimes involves fairly complex checking and tree walking.

As one example this complexity, review [`AccessibilityTable:isDataTable()` in the WebKit source](https://github.com/WebKit/webkit/blob/master/Source/WebCore/accessibility/AccessibilityTable.cpp#L112). 

In addition to hierarchical validation and well-formedness (`<tables>` do not get the table _role_ unless they have a valid grid of rows and cells), this method checks for a variety of style-based heuristics such as borders and zebra-striped row colors. It's determining layout-vs-data table based on the style characteristics, which then determine the role, and would only then affect the cascade for a `:role()` selector.

**Comment by SelenIT** [comment]
I have a strong feeling that distinguishing data tables from layout table is a special, largely historical, exceptional case, that should be not generalized as a rule. Tables were used for layout when there were effectively no alternative. Tables are probably the only HTML elements misused this way because of their default styling. There never were such thing as "Layout lists" or something. The "HTML is meaning, CSS is presentation" mantra is one of the first things that modern web developers learn. Many web developers try hard to make their markup meaningful for non-visual output (e.g. wrapping navigation links into list items) while presenting it visually as design requires (e.g. as a horizontal flexbox "ribbon"). When browsers deduce from this that the author "[didn't mean](https://www.scottohara.me/blog/2019/01/12/lists-and-safari.html)" a list there, it feels like a huge disappointment with the very idea of the semantic markup (at least, I felt so).

I admit that there definitely might be other special cases where styling distorts the initial meaning of elements for sighted users and ATs have to convey this distortion, but they should probably be investigated and specified individually. The general rule should be that CSS doesn't affect the implied meaning of the content unless there is an absolute and consensus-approved necessity to do so. Similarly to what we have already decided in [#2355](2355#issuecomment-375084853).

**Comment by Crissov** [comment]
So, `:role()` would be a bit like `:lang()` in that it can be inherited or inferred, while `[role]` or `[aria-role]` cannot. A pseudo-class seems appropriate then, otherwise I would have suggested special syntax as for IDs `#foo` and classes `.bar`, e. g. `@baz` or `%baz`.

**Comment by bkardell** [comment]
@Crissov can you explain that comment a little more?

**Comment by bkardell** [comment]
@SelenIT - I think @alice or @cookiecrook can speak better to how unique that particular example is, but there's definitely additional relationships and complexity which I'm unsure even happens at the right stage right now to be considered in a selector.  I believe some roles may actually rely on CSS application in order to determine their computed role, which would be problematic.  I'm not suggesting that this isn't a thing that shouldn't be supported, and if we were designing all of the things fresh, this is a thing I would really want to keep in mind - I just wonder if we might not need to think about this a lot more and maybe work on some efforts to detangle - or maybe thinks about revise the ask somehow in a way that could be achievable but not just add confusion for developers about one more hard to understand kind of value?

**Comment by Crissov** [comment]
@bkardell At first, I wondered how `:role(foo)` would be any different from `[role=foo]`, but then I realized that attribute selectors work on explicitly specified values while pseudo-classes can work on implicit values. Also, IDs and classes are not predefined by an external body while ARIA roles and BCP47 language codes are.

**Comment by ExE-Boss** [comment]
Also, adding completely new special syntax might not work so well with https://github.com/w3c/csswg-drafts/issues/3264.

**Comment by cookiecrook** [comment]
@SelenIT wrote:
> I have a strong feeling that distinguishing data tables from layout table is a special, largely historical, exceptional case, that should be not generalized as a rule.

Here's the same concept for layout lists (a.k.a. "list-itis") versus semantic lists:  
[`AccessibilityList::determineAccessibilityRole()` in the WebKit source](https://github.com/WebKit/webkit/blob/master/Source/WebCore/accessibility/AccessibilityList.cpp#L127)

"div-itis" is also a thing, though thankfully that doesn't get an important role by default. Other element types that my get this treatment at some point include `<article>`, `<header>`, `<footer>`, etc.

**Comment by cookiecrook** [comment]
@SelenIT 
> When browsers deduce from this that the author "[didn't mean](https://www.scottohara.me/blog/2019/01/12/lists-and-safari.html)" a list there, it feels like a huge disappointment with the very idea of the semantic markup (at least, I felt so).

Not to diverge too far off topic, but threads on that are here:
https://twitter.com/pgrucza/status/1083835348671700992
https://twitter.com/cookiecrook/status/1084141791580905472

**Comment by SelenIT** [comment]
> the same concept for layout lists (a.k.a. "list-itis")

@cookiecrook, I agree with your point about Priority of Constituencies in the twitter discussions, but I'd like to point out to what I believe is a significant difference between misusing tables for layout and displaying lists with `list-style-type: none`. Tables were used for layout _because of their display,_ ignoring the semantics. Lists were used for things that don't look list-like _because of their semantics,_  to convey that the following items belong to the same set, and specifically to enable non-visual aids like the count of the items. This seems to be the opposite to the "layout tables" concept. 

Yes, visual representation of these items can differ from the typical bulleted/numbered list in text, but usually not to the extent that the sighted user stops to interpret them as e.g. "list of 5 horizontal navigation links" or "portion of the 15 items of the catalog list sorted by price". And I'm not sure that hiding this intended meaning that authors _did_ put into the markup from the non-visual users serves the users' priority, rather than the _purely theoretical_ concept that "screenreader should read only what is explicitly on the screen".

Surely, as an author, I might be biased about this:). Is there some evidence that presenting lists without explicit list styles as lists to screenreader users harms their UX?

**Comment by andyearnshaw** [comment]
I discovered this issue after thinking about how to write some generic code for a web component that will use an event handler to find the nearest ancestor with a specific role and then get some more information about it. It looks like there's a current limitation, where a custom element has role set via `ElementInternals.role`, which means you cannot rely on a CSS selector to find that custom element by its role. For example,

```
myCustomTable.addEventHandler('click', ({ target }) => {
  // This won't find an element whose role is set with ElementInternals.role
  const clickedCell = target.closest('td, [role="cell"]');

  if (clickedCell) {
    // There's a separate, but similar problem here that AOM should address
    const [x, y] = getColumnAndRowIndex(clickedCell);
    handleCellClick(x, y);
  }
});
```

Not only would `:role()` simplify the selector, it would cover this current limitation.

**Comment by cookiecrook** [comment]
> > the same concept for layout lists (a.k.a. "list-itis")
> 
> @cookiecrook, thank you for these links. I completely agree with the Priority of Constituencies principle you cite. Hovewer, my point that there is _no such thing as layout lists_ still stands. The term "layout lists" [is a misnomer](https://yatil.net/blog/much-ado-about-no-lists#fn-1). 

Here's a [more complete thread of the background on layout lists](https://twitter.com/cookiecrook/status/1337226933822603270?lang=en) and why I would expect WebKit's behavior to remain. 

Regardless, the lists example tangential to the point of this issue. There are other contextual role roles in [ARIA](https://github.com/web-platform-tests/wpt/pull/42412) and [HTML-AAM](https://wpt.fyi/results/html-aam/roles-contextual.html?label=master&label=experimental&aligned&q=label%3Aaccessibility). While I agree that it'd be _possible_ to extract some or all of this logic from the accessibility runtime and insert it into the mainstream runtime, it's a non-trivial amount of work that, so far, no implementor I'm aware of has agreed to take on.

**Comment by LeaVerou** [comment]
To echo my comment from #10222, I’m not sure what the use cases are here. Is the idea that authors would switch their default styles from using the kinds of long selectors discussed in #2296 and would use `:role(textbox)` and `:role(button)` instead? If so, I’m unsure if that’s a good idea as it seems to mix style and behavior — in most web apps there are buttons that look like buttons, and buttons that don't look anything like buttons, especially given the lack of good ARIA roles to describe today’s UIs. E.g. even interactive visualizations where you hover over a e.g. a pie chart wedge or scatterplot circle to see more info about that data have to use `role=button` because there’s nothing better.

**Comment by Crissov** [comment]
The apparent need for additional roles in ARIA does not negate the usefulness of a role selector in CSS. If that’s an issue in practice it should be brought up in the proper place. 

____
(The following is a draft comment from a while ago which I rediscovered just now and shall publish without any reevaluation:)

I’d like to point out that someone could create an external specification that assigned ARIA implicit default roles to elements in very different markup languages, e.g. ones without obvious element names like Markdown, thereby paving the path of adding direct CSS styling support to them.

**Comment by svgeesus** [comment]
Selectors 4 is being stabilized to move to Recommendation. Proposed new features are being re-tagged for Selectors-5.
